Add new extract dev command to retrieve old versions of formulae

alebcay opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

This is my first time contributing to `Homebrew/brew` and in fact my first time ever writing Ruby outside of the Homebrew and Homebrew-Cask DSLs. So please bear with me - there's probably parts of this PR that are unidiomatic if not downright ugly. I may have re-invented the wheel a few times too.

This PR introduces the start of a `brew extract` command, intended to address #4055. It extracts the formula file for a specified formula that matches a given version and makes a copy of the file into the `Formula` subdirectory of a specified tap.

The syntax is along the lines of `brew extract <formula> <version> <tap>` and the new file is placed at `<tap>/Formula/<formula>@<version>.rb`. If the file already exists at that path, it will not overwrite the file unless `--force` is specified.

In it's current state, something is wrong with how I'm getting the "next-most recent" commit/revision to be considered. After a few iterations of searching via an `until` loop, the revisions are all empty, although if I write out a single iteration from the last good revision that it gave me, it'll still return me the proper next commit hash. It's almost as if the Git-querying part of the code stops responding after a certain amount of iterations in the loop but works perfectly fine outside of it. It's possible that I'm just overlooking some trivial Ruby-ism right now that's causing this, of course.

-----

As a demonstration, the following command will work properly:
<details>

```
• brew extract --debug libogg 1.3.2 homebrew/core
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.3 from revision HEAD against desired 1.3.2
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.3 from revision ddb386b against desired 1.3.2
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.3 from revision f64a373 against desired 1.3.2
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.2 from revision 16ebe5f against desired 1.3.2
==> Writing formula for libogg from 16ebe5f to /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg@1.3.2.rb
```

</details>

but the following will not:
<details>

```
• brew extract --debug libogg 1.3.1 homebrew/core
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.3 from revision HEAD against desired 1.3.1
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.3 from revision ddb386b against desired 1.3.1
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.3 from revision f64a373 against desired 1.3.1
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.2 from revision 16ebe5f against desired 1.3.1
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.2 from revision b100d2d against desired 1.3.1
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.2 from revision 23d09e4 against desired 1.3.1
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.2 from revision 8b378fd against desired 1.3.1
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.2 from revision 5ac836a against desired 1.3.1
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.2 from revision 893cd40 against desired 1.3.1
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.2 from revision 1fda5d2 against desired 1.3.1
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.2 from revision f4bd9ce against desired 1.3.1
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> formulae require at least a URL in libogg at revision 61b2307
/usr/local/Homebrew/Library/Homebrew/formula.rb:244:in `determine_active_spec'
/usr/local/Homebrew/Library/Homebrew/formula.rb:197:in `initialize'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:82:in `new'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:82:in `get_formula'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:327:in `from_contents'
/usr/local/Homebrew/Library/Homebrew/formula_versions.rb:40:in `block in formula_at_revision'
/usr/local/Homebrew/Library/Homebrew/utils.rb:425:in `nostdout'
/usr/local/Homebrew/Library/Homebrew/formula_versions.rb:40:in `formula_at_revision'
/usr/local/Homebrew/Library/Homebrew/cmd/extract.rb:37:in `extract'
/usr/local/Homebrew/Library/Homebrew/brew.rb:87:in `<main>'
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaContentsLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libogg.rb
==> Trying 1.3.3 from revision  against desired 1.3.1
Error: Could not find libogg 1.3.1.
/usr/local/Homebrew/Library/Homebrew/utils.rb:60:in `exit'
/usr/local/Homebrew/Library/Homebrew/utils.rb:60:in `odie'
/usr/local/Homebrew/Library/Homebrew/cmd/extract.rb:39:in `extract'
/usr/local/Homebrew/Library/Homebrew/brew.rb:87:in `<main>'
```
</details>

-----

Other functionality that I've considered but am not sure whether or not would be a good idea to implement:
- Allowing user to pass a git ref instead of a version (e.g. `brew extract foo 1d9f4c3 homebrew/custom`)
- Searching through legacy homebrew repo(s) or user-specified taps as well (instead of just `Homebrew/homebrew-core`)
- `--stdout` flag or something like that to just dump the formula on the terminal instead of to a file

Fixes #4055

MikeMcQuaid commented on Library/Homebrew/cmd/extract.rb:
```diff
@@ -0,0 +1,53 @@
+#:    not installed yet, attempts to install/clone the tap before continuing.
+#:
+#:    If the file at <tap>/Formula/<formula>@<version>.rb already exists,
+#:    it will not be overwritten unless `--force` is specified.
```
Please document the `--force` option.

MikeMcQuaid commented on Library/Homebrew/cmd/extract.rb:
```diff
@@ -0,0 +1,53 @@
+module Homebrew
+  module_function
+
+  def extract
```
Please use `Homebrew::CLI::Parser.parse do` (like `audit.rb`)

MikeMcQuaid commented on Library/Homebrew/cmd/extract.rb:
```diff
@@ -0,0 +1,53 @@
+  module_function
+
+  def extract
+    formula = Formulary.factory(ARGV.named[0])
```
`ARGV.formulae.first` and needs handling for if a no arguments are provided.

alebcay replied:
Replacing this line with `formula = ARGV.formulae.first` gives the result `Error: No available formula with the name "<version>"`.

MikeMcQuaid replied:
No worries, my mistake; didn't think about how you're using the other arguments. `ARGV.name.first` is sufficient (although you should check the value isn't `nil` first).

MikeMcQuaid commented on Library/Homebrew/cmd/extract.rb:
```diff
@@ -0,0 +1,53 @@
+
+  def extract
+    formula = Formulary.factory(ARGV.named[0])
+    version = ARGV.named[1]
```
Needs handling if no `version` is provided.

MikeMcQuaid commented on Library/Homebrew/cmd/extract.rb:
```diff
@@ -0,0 +1,53 @@
+  def extract
+    formula = Formulary.factory(ARGV.named[0])
+    version = ARGV.named[1]
+    dest = Tap.fetch(ARGV.named[2])
```
Needs handling if no `dest` is provided.

MikeMcQuaid replied:
`destination_tap` would be a nicer name.

MikeMcQuaid commented on Library/Homebrew/cmd/extract.rb:
```diff
@@ -0,0 +1,53 @@
+          To overwrite it and continue anyways, run `brew extract #{formula} #{version} #{dest.name} --force`.
+        EOS
+      end
+      ohai "Clobbering existing formula at #{path}" if ARGV.debug?
```
`Overwriting existing formula at`

MikeMcQuaid commented on Library/Homebrew/cmd/extract.rb:
```diff
@@ -0,0 +1,53 @@
+
+    rev = "HEAD"
+    version_resolver = FormulaVersions.new(formula)
+    rev = Git.last_revision_commit_of_file(formula.path.parent.parent, formula.path, before_commit: "#{rev}~1") until version_resolver.formula_at_revision(rev) { |f| version_matches?(f, version, rev) || rev.empty? }
```
Use `until do` block form so this line is shorter

MikeMcQuaid commented on Library/Homebrew/cmd/extract.rb:
```diff
@@ -0,0 +1,53 @@
+    version_resolver = FormulaVersions.new(formula)
+    rev = Git.last_revision_commit_of_file(formula.path.parent.parent, formula.path, before_commit: "#{rev}~1") until version_resolver.formula_at_revision(rev) { |f| version_matches?(f, version, rev) || rev.empty? }
+
+    odie "Could not find #{formula} #{version}." if rev.empty?
```
End with `!`

MikeMcQuaid commented on Library/Homebrew/cmd/extract.rb:
```diff
@@ -0,0 +1,53 @@
+    ohai "Writing formula for #{formula} from #{rev} to #{path}"
+
+    # The class name has to be renamed to match the new filename, e.g. Foo version 1.2.3 becomes FooAT123 and resides in Foo@1.2.3.rb.
+    path.write result.gsub("class #{formula.name.capitalize} < Formula", "class #{formula.name.capitalize}AT#{version.gsub(/[^0-9a-z ]/i, "")} < Formula")
```
use `result.gsub!` and separate each `gsub` and `write` onto a separate line.

MikeMcQuaid commented:
Run `brew man` and commit the results. Would also be good to have an integration test for this (like others in `Library/Homebrew/test/cmd/`)

commitay commented:
Should this be a dev command like create, edit and tap-new?

MikeMcQuaid commented:
> Should this be a dev command like create, edit and tap-new?

@commitay Yep 👍

alebcay commented:
Reworked the code based on the nitpicks above. A `UsageError` is now raised if the wrong number of (named) arguments are provided (expect three when writing to tap, two when writing to stdout). A regular `odie` is thrown if the user tries to specify a tap and also use `--stdout` at the same time. This guards and simplifies me from having to `... unless <var>.nil?` the three arguments separately.

Also generated and committed the manpages for the command. Tests still need to be written after we actually get the command working. The issue I originally stated regarding the strange behavior of `last_revision_commit_of_file(repo, file, before_commit: nil)` just not returning anything after a certain number of iterations is still happening.

alebcay commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,75 @@
+    end
+
+    odie "Cannot use a tap and --stdout at the same time!" if ARGV.named.length == 3 && args.stdout?
+    raise UsageError unless (ARGV.named.length == 3 && !args.stdout?) || (ARGV.named.length == 2 && args.stdout?)
```
How would be a good way of going about figuring out which one is missing? i.e. how would I tell that `brew extract libogg 1.3.2` is missing a tap compared to `brew extract 1.3.2 homebrew/custom` is missing a formula? Would this be easier if we explicitly specified each arg with a switch (like `brew extract --formula libogg --version 1.3.2 --tap homebrew/custom`)?

MikeMcQuaid replied:
Yeh, you could perhaps make the tap the second argument rather than the third or something like `--tap=` and maybe even `--version=`.

alebcay commented:
`stdout` option is now removed (a tap must be specified via `--tap=`). The version is now optional (can be specified via `--version=`) and the most recent is taken if no version is specified. If the specified tap happens to be `homebrew/core`, it also bails. Message for re-running with `--force` now spans another line and should be within 80 chars. Lastly, cleaned up the class name generation part. We now perform just one `gsub!` in order to actually replace the first line of the file.

Still two major issues with this code as it stands:
- Does not work at all with formulae that have been deleted. I think we'll have to stop relying on `FormulaVersions` and rely solely on `Git.last_revision_commit_of_file(repo, file, before_commit: nil)` and `Git.last_revision_of_file(repo, file, before_commit: nil)` to get those. It would be great if there would be some easy way to extract and compare the version from the file contents. I tried this and the only way I could think of was to put the contents of the last revision into a temp file and have Formulary read it from a path and see if it's the right version before deciding whether or not to dig further in the history. Hopefully there's a less kludgy way to go about this that I just don't know about.
- Still have issues with `Git.last_revision_commit_of_file(repo, file, before_commit: nil)` deciding to stop working at times inside of loops.

MikeMcQuaid commented:
Thanks @alebcay! I'll make a note to have a look at the Git stuff when I get a chance.

alebcay commented:
I may have made some headway on that Git issue:

```
[34] brew(main)> resolver.formula_at_revision("1fda5d2") { |f| f.version }
=> #<Version::FromURL:0x00007f9100a747f0 @version="1.3.2">
[35] brew(main)> resolver.formula_at_revision("9274aca") { |f| f.version }
=> nil
```

([`9274aca`](https://github.com/Homebrew/homebrew-core/commit/9274aca) is the next most recent commit after [`1fda5d2`](https://github.com/Homebrew/homebrew-core/commit/1fda5d2) to have modified the formula in question - in this case, `libogg`)

At some point in the revision history, `FormulaVersions.formula_at_revision(rev)` is unable to get the version from the formula anymore.

-----

As another example, `automake` only has one "good" revision ([`1e1a871`](https://github.com/Homebrew/homebrew-core/commit/1e1a871)) before we have the same issue ([`61dae90`](https://github.com/Homebrew/homebrew-core/commit/61dae90)):

```
[10] brew(main)> resolver.formula_at_revision("1e1a871") { |f| f.version }
=> #<Version::FromURL:0x00007fbdbb51dff8 @version="1.16.1">
[11] brew(main)> resolver.formula_at_revision("61dae90") { |f| f.version }
=> nil
```

MikeMcQuaid commented:
@alebcay It may be the older versions are invalid formulae i.e. use methods that have been since removed?

alebcay commented:
That's what I thought at first too, but the difference between working and non-working revisions of formulae is minute - the `sha256` of the bottle being updated is somehow able to break things (in `libogg`'s example).

Is there some kind of logic we could use to extract the URL from the formula file without constructing a proper Formula object, or better yet, easily extract the version from the URL from said formula? The first rather kludgy idea was to look for lines beginning with `url` and get the URL that way, but it's not very versatile since formulae can have `resource` blocks with URLs too.

alebcay commented:
Another progress update - I've gotten the problems out of the way now, mostly by ignoring missing methods (`sha1`, `md5`) in `Module` and `BottleSpecification`. Right now, this code can go as far back as [this commit](https://github.com/Homebrew/homebrew-core/commit/4139352) before running into issues. Any ideas on how to reconcile/handle those older formulae, or are they just so old that I shouldn't bother?

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+  end
+
+  prepend Compat
+end
```
Definitely don't want to have to maintain all the code above indefinitely. I'd string out the `ARGV.debug?` stuff and just make the bad values a no-op or convert them to a good value.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+    # If some other number of args are given, provide generic usage information
+    raise UsageError if ARGV.named.length != 1
+
+    odie "The tap to which the formula is extracted must be specified!" if args.tap.nil?
```
I wonder if it's worth making it a non-flagged argument (e.g. it's just the second argument)?

alebcay replied:
I'll go in this direction. Just feels wrong making a flagged argument mandatory. In this case, we won't raise a separate `FormulaUnspecifiedError`, because there's no (easy) way to check if a missing argument is the missing formula or if it's the tap that needs to be specified. Instead, just a generic `UsageError` will be raised if we don't get exactly 2 named args.

MikeMcQuaid replied:
> Just feels wrong making a flagged argument mandatory

MikeMcQuaid replied:
> Just feels wrong making a flagged argument mandatory

Agreed.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+    raise FormulaUnspecifiedError if ARGV.named.empty?
+
+    # If some other number of args are given, provide generic usage information
+    raise UsageError if ARGV.named.length != 1
```
What's the output look like for this?

alebcay replied:
```
brew extract [--force] formula --tap=tap [--version=version]:
    Looks through repository history to find the version of formula and
    creates a copy in tap/Formula/formula@version.rb. If the tap is
    not installed yet, attempts to install/clone the tap before continuing.
    A tap must be passed through --tap in order for extract to work.

    If --force is passed, the file at the destination will be overwritten
    if it already exists. Otherwise, existing files will be preserved.

    If an argument is passed through --version, version of formula
    will be extracted and placed in the destination tap. Otherwise, the most
    recent version that can be found will be used.

Error: Invalid usage
```

MikeMcQuaid replied:
Great.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+    end
+
+    # If no formula args are given, ask specifically for a formula to be specified
+    raise FormulaUnspecifiedError if ARGV.named.empty?
```
What's the output look like for this?

alebcay replied:
```
brew extract [--force] formula --tap=tap [--version=version]:
    Looks through repository history to find the version of formula and
    creates a copy in tap/Formula/formula@version.rb. If the tap is
    not installed yet, attempts to install/clone the tap before continuing.
    A tap must be passed through --tap in order for extract to work.

    If --force is passed, the file at the destination will be overwritten
    if it already exists. Otherwise, existing files will be preserved.

    If an argument is passed through --version, version of formula
    will be extracted and placed in the destination tap. Otherwise, the most
    recent version that can be found will be used.

Error: Invalid usage: This command requires a formula argument
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+      name = formula.name
+      repo = formula.path.parent.parent
+      file = formula.path
+    rescue FormulaUnavailableError => e
```
Why bother with the non-rescue version of this code?

MikeMcQuaid replied:
i.e. what advantage does the code above the `rescue` have over the code below?

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+      core = Tap.fetch("homebrew/core")
+      name = ARGV.named.first.downcase
+      repo = core.path
+      file = core.path.join("Formula", "#{name}.rb")
```
Prefer `/` over `.join`

alebcay replied:
Would `core.path.join("Formula/#{name}.rb")` be all right? Or just remove it entirely, but create Pathname object around the whole thing (`Pathname.new("#{core.path}/Formula/#{name}.rb"`)? At some point, it'll need to be converted into a `Pathname` because `Git.last_revision_of_file(repo, file, before_commit: rev)` expects arguments as `Pathname` objects, not `String`.

reitermarkus replied:
`core.path/"Formula/#{name}.rb"`

MikeMcQuaid replied:
(which returns a `Pathname`)

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+    destination_tap = Tap.fetch(args.tap)
+    destination_tap.install unless destination_tap.installed?
+
+    odie "Cannot extract formula to homebrew/core!" if destination_tap.name == "homebrew/core"
```
Would be good to put this up with the other argument handling.

reitermarkus replied:
`destination_tap.core_tap?`

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+
+    odie "Cannot extract formula to homebrew/core!" if destination_tap.name == "homebrew/core"
+
+    if args.version.nil?
```
`if args.version` and flip it?

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+
+  # @private
+  def formula_at_revision(repo, name, file, rev)
+    return nil if rev.empty?
```
`return if rev.empty?` returns `nil` by default

reitermarkus commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+      file = formula.path
+    rescue FormulaUnavailableError => e
+      opoo "'#{ARGV.named.first}' does not currently exist in the core tap" if ARGV.debug?
+      core = Tap.fetch("homebrew/core")
```
`CoreTap.instance`

reitermarkus commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+  end
+end
+
+class Module
```
Dont' monkey-patch this.

MikeMcQuaid replied:
@reitermarkus Do you have an alternate suggestion to handling old formulae?

reitermarkus replied:
We can use refinements to limit the monkey-patch to a certain scope.

MikeMcQuaid replied:
Seems like a good idea 👍

reitermarkus commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+require "formulary"
+require "tap"
+
+class BottleSpecification
```
Dont' monkey-patch this.

MikeMcQuaid replied:
@reitermarkus Are you OK with the current implementation given the discussion?

MikeMcQuaid replied:
@reitermarkus Ping.

reitermarkus commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,153 @@
+  end
+end
+
+class DependencyCollector
```
Dont' monkey-patch this.

alebcay commented:
Made a feeble attempt at using Ruby refinements instead of just monkeypatching. In this state they don't work (it complains about `sha1` etc. not found on older formulae, which is what the monkeypatches fixed). I'm probably overlooking some nuance of refinements, but figured I'd update my progress here instead of just banging my head on the wall.

reitermarkus commented:
Oh right! The refinements are scoped, so it won't work if a missing method on a `BottleSpecification` gets called from another file. 🤦‍♂️

alebcay commented:
Would it be all right just to do a monkey-patch like before then?

alebcay commented:
All right, another batch of changes! This version now works on formulae as old as the `homebrew-core` repo itself (i.e. all the way back to the initial commit in the repo). I'll also get around to starting on writing up spec/tests.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,131 @@
+
+class BottleSpecification
+  def method_missing(m, *_args, &_block)
+    # no-op
```
Can skip all the no-op comments.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,131 @@
+require "tap"
+
+class BottleSpecification
+  def method_missing(m, *_args, &_block)
```
Can just do `method_missing(*)` if you're not using any parameters

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,131 @@
+    when :java       then JavaRequirement.new(tags)
+    when :osxfuse    then OsxfuseRequirement.new(tags)
+    when :tuntap     then TuntapRequirement.new(tags)
+    when :ld64       then ld64_dep_if_needed(tags)
```
Make all these values a no-op instead.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,131 @@
+
+  module Compat
+    def parse_string_spec(spec, tags)
+      opoo "'depends_on ... => :run' is disabled. There is no replacement." if tags.include?(:run) && ARGV.debug?
```
Can just likely make this a `super`

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,131 @@
+
+    core = CoreTap.instance
+    name = ARGV.named.first.downcase
+    repo = core.path
```
`repo = CoreTap.instance.path`

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,131 @@
+    odie "Cannot extract formula to homebrew/core!" if destination_tap.core_tap?
+    destination_tap.install unless destination_tap.installed?
+
+    core = CoreTap.instance
```
Can remove this

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,131 @@
+    core = CoreTap.instance
+    name = ARGV.named.first.downcase
+    repo = core.path
+    file = core.path/"Formula/#{name}.rb"
```
`file = repo/"Formula/#{name}.rb"` although it's worth noting not all repos have this structure (the `Formula` directory is optional)

alebcay replied:
I'm aware that the `Formula` dir isn't required, although I figured that it would be fine since we're only reading from the core tap, which _does_ have the directory. Should I consider generalizing this logic a bit so that reading from other taps (which might not have the `Formula` dir) is possible down the road?

MikeMcQuaid replied:
@alebcay Yeh, it's fine, might want to note it as a comment though just so others looking at the code are aware 👍

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,131 @@
+      result = Git.last_revision_of_file(repo, file)
+    end
+
+    # The class name has to be renamed to match the new filename, e.g. Foo version 1.2.3 becomes FooAT123 and resides in Foo@1.2.3.rb.
```
This line needs wrapped, it's a bit too long.

MikeMcQuaid commented:
@alebcay Can you rebase this on `master`? Thanks! Not sure why the tests aren't running but that'll fix it.

alebcay commented:
Done, looks like the tests are running now.

MikeMcQuaid commented on Library/Homebrew/test/dev-cmd/extract_spec.rb:
```diff
@@ -0,0 +1,13 @@
+    path = Tap::TAP_DIRECTORY/"homebrew/homebrew-foo"
+    (path/"Formula").mkpath
+    target = Tap.from_path(path)
+    formula_file = setup_test_formula "foo"
```
You might need to actually create a Git repository here and a few revisions?

MikeMcQuaid commented:
@alebcay `brew style` is failing unfortunately.

alebcay commented:
The spec/integration test is now written. Probably not idiomatic, but functional and actually tests stuff. There are still two style violations, both regarding that `super` should be used in `def method_missing(*)`, while I think we came to the agreement that it should just be a no-op, lest I do some kludging to squeeze in a `super` call in some unreachable code somehow.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,126 @@
+require "formulary"
+require "tap"
+
+class BottleSpecification
```
@alebcay Use `# rubocop:disable Style/MethodMissingSuper` above this line to fix `brew style`. Use `# rubocop:enable Style/MethodMissingSuper` above `module Homebrew` to re-enable it there. Thanks!

MikeMcQuaid commented on Library/Homebrew/test/dev-cmd/extract_spec.rb:
```diff
@@ -0,0 +1,235 @@
+describe "brew extract", :integration_test do
```
Integration tests are comparatively really slow. As a result I think it's sufficient to have a single test that verifies the happy path. If you wish to add more tests consider refactoring some logic out of `cmd` and into something that can be unit tested (but I think a single test is fine here for now).

MikeMcQuaid replied:
This was addressed 👍

alebcay commented:
All style issues should be fixed now. I heavily reduced/combined the integration tests. The corner cases would indeed probably do better in a unit test. I have considered refactoring parts of the logic in this feature out of `dev-cmd` (into somewhere else in the core tree), but figured that it really isn't much use to anyone/anything else at the moment.

MikeMcQuaid commented:
@alebcay The code looks good! Feedback from having run it and some suggested, very small final changes.

- `brew extract inkscape homebrew/bundle` pauses for a long time before failing with `Error: inkscape: undefined method `sha1' for #<Resource::PatchResource:0x000000010207cc88>`. That error might want to be addressed but I'd suggest at least making it print a message (with `ohai`) indicating that it's e.g. searching the Git history?
- `brew extract wget homebrew/bundle --version=1.16.1` worked as expected but `brew install wget@1.16.1` failed due to `undefined method `sha1' for #<Class:0x00000001011dafe0>`. I wonder if it's worth stripping the bottle blocks or e.g. `sha1` etc.? I think it's fine to say "good idea for a follow up PR (by someone else)".
- `brew extract wget homebrew/bundle --version=1.19.2` and `brew install wget@1.19.2` worked as expected.
- I think it's worth adding a reference to this command in https://github.com/Homebrew/brew/blob/master/docs/Versions.md. That's exactly the type of resource we point people to to justify why we won't provide every version; it'd be perfect to recommend it there.

alebcay commented:
> `brew extract inkscape homebrew/bundle` pauses for a long time before failing

Added a progress message. The failure was because we needed to monkey-patch `method_missing(*)` in `PatchResource`. I just did the monkey-patch on `Resource` instead to cover any other corner cases where `resource` checksums were `sha1` instead of `sha256`. If you can think of anywhere else in the codebase where we've been using checksums and have upgraded from `sha1`/`md5` to `sha256`, they also probably need a monkey-patch.

> `brew extract wget homebrew/bundle --version=1.16.1` worked as expected but `brew install wget@1.16.1 failed`

I suppose a question to ask is - how usable are we expecting formula that this churns out to be? We could certainly strip `sha1`/other missing blocks, but edits would have to be made to add a `sha256` block too, right? I think it's beyond the scope of this command (and definitely outside of the scope of this PR) to have the `extract` command not only retrieve, but also format old Homebrew stanzas/replace them with modern alternatives. At least, I was under the assumption that the formulae that are churned out here are "use at your own risk" or "needs some fixing up" by the end user. I'm already a bit concerned that we're packing a lot of logic into a single command, and if we were to add this other stuff, I think it may warrant refactoring some of the stuff out of `dev-cmd`.

I'll get around to putting a word in the docs about the command.

> For unit-testing, the monkey-patches should be reversible.

Where would this thing go? In the `extract_spec.rb`, and I'd just call it?

reitermarkus commented:
> Where would this thing go?

Loading the formula would happen inside `with_monkey_patch`, so the monkey-patch is only active when loading an old formula.

reitermarkus commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,160 @@
+require "formulary"
+require "tap"
+
+def with_monkey_patch
```
Also add the `DependencyCollector` monkey-patch here.

reitermarkus commented on Library/Homebrew/dev-cmd/extract.rb:
```diff
@@ -0,0 +1,160 @@
+  def parse_symbol_spec(*); end
+
+  module Compat
+    def parse_string_spec(spec, tags)
```
What is this doing that `def parse_symbol_spec(*); end` doesn't?

alebcay replied:
I was trying to monkey-patch over [this `compat` definition of `parse_symbol_spec](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/compat/dependency_collector.rb); the [`super` one](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/dependency_collector.rb) works fine as far as I can tell.

Of course, we could just clobber over it with a no-op and I think we'd be able to move on just fine.

alebcay commented:
Done, although tests are failing because the new monkey-patch doesn't work now with `HOMEBREW_NO_COMPAT=1`:

```
Error: uninitialized constant DependencyCollector::Compat
```

Should I just guard the monkey-patch to check the env variable, or just `require` the relevant file?

reitermarkus commented:
You could check for `defined?(DependencyCollector::Compat)`.

MikeMcQuaid commented:
Thanks for all the hard, great work here @alebcay!

> To be clear, I don't mind stripping old stuff out of the resulting formulae, but I just don't feel that it's much of an improvement if we don't replace them with their modern equivalents - and that part is considerably more complicated.

Agreed this can be done in a future PR if someone needs it (and not necessarily by you).

> I'll get around to putting a word in the docs about the command.

That'd be great, thanks!
